app/models/manageiq/providers/amazon/inventory/parser/cloud_manager.rb

```diff
@@ -206,7 +215,7 @@ def instance_hardware(persister_instance, instance, flavor)
       :cpu_total_cores      => flavor[:cpus],
       :memory_mb            => flavor[:memory] / 1.megabyte,
       :disk_capacity        => flavor[:ephemeral_disk_size],
-      :guest_os             => persister.hardwares.lazy_find(instance['image_id'], :key => :guest_os),
+      :guest_os             => persister.hardwares.lazy_find(persister.miq_templates.lazy_find(instance['image_id']), :key => :guest_os),
```

Comment:
@agrare  so, I am still keeping the old way that we do not need to define a key, if the manager_ref is size 1

so
```ruby
persister.hardwares.lazy_find(persister.miq_templates.lazy_find(instance['image_id']))
```
is equivalent to:
```ruby
persister.hardwares.lazy_find(:vm_or_template => persister.miq_templates.lazy_find(:ems_ref => instance['image_id']))
```

and instead of lazy_find, we can use the built object
```
persister.hardwares.lazy_find(persister.miq_templates.find_or_build(instance['image_id']))
```

Or would you prefer just the hash syntax?

Reply:
cc @cben what do you think ^ , I am now working on building Db queries automatically, using these (also for secondary indexes)

Reply:
OK, VmOrTemplate has_one hardware, hardwares manager_ref is [:vm_or_template].  Now I can read this ;-)

- Supporting both single value and hash makes sense, consistent with existing interface, though for hashes I prefer using `lazy_find_by` and eventually phasing out `lazy_find(hash)`.
- Allowing `find_or_build` too makes sense.  A major convenience of inventory objects is making actual records, records-to-be, and lazy refs all interchangeable.

`:key` with hash syntax gets a little tricky.  I don't think we'd want it mixed on same level:
```
persister.hardwares.lazy_find_by(
  :vm_or_template => persister.miq_templates.lazy_find(:ems_ref => instance['image_id']), 
  :key            => :guest_os  # confusing, this is not a column
)
```
but I think requiring a sub-hash would be fine:
```
persister.hardwares.lazy_find_by(
  {
    :vm_or_template => persister.miq_templates.lazy_find(:ems_ref => instance['image_id']),
    # would be a better example with a composite manager_ref of several columns
  }
  :key => :guest_os
)
```

Reply:
@cben right, so this kind of sucks with kwargs

persister.hardwares.lazy_find_by(
  :vm_or_template => persister.miq_templates.lazy_find(:ems_ref => instance['image_id']), 
  :key            => :guest_os  # confusing, this is not a column
)

Would be taken as the 1 argument is Hash, not sure how to enforce and assert to pass a subhash.

Also, then plan was to get rid of _by methods, right now those are just calling the normal methods.  (lazy_find_by just calls lazy_find). So maybe we can keep the _by variant after all? With smart asserts, we could avoid it being confusing like it is now. @agrare what do you think?